### ISSUE TITLE
`numpy.distutils` is deprecated in NumPy 1.23 and issues with `setuptools>=65`.

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#8355]
### Introduction:
In trying to get to the root cause of this backtrace: https://github.com/numba/numba/pull/8352#issuecomment-1215020818 some things of note:
1. `setuptools >= 65` has removed the `msvc` compiler exports that Numba was using via NumPy (see https://github.com/numba/numba/pull/8352#issuecomment-1215026200)
2. NumPy 1.23 has deprecated its `numpy.distutils` module with removal scheduled for Python >= 3.12. To trigger the deprecation notice, using NumPy 1.23, run:
   ```
   python -c "import numpy.distutils.misc_util"
   ```
3. The change in 1. leads to NumPy's `distutils` compilation chain being broken. MWR that doesn't involve Numba at all:
   ```python
   import numpy.distutils.misc_util
   from cffi import FFI
   ffi = FFI()
   ffi.set_source('tmp', "int foo(void) { return 1; }")
   ffi.compile()
   ```
   this triggers the backtrace noted above.

I think that the impact of the `setuptools` change is that `numba.distutils` for _all_ NumPy versions will be broken.

-----------
### Numba is impacted by this as follows:

* This manifests as an issue in Numba's build as the extension modules creation in `setup.py` has this: https://github.com/numba/numba/blob/f5af7b64b98bb745c78cf16913b16acb34017dd2/setup.py#L136-L143

* This manifests as issues in Numba's testing as `numba.tests.support`, which is imported in pretty much every single Numba test case to get the custom `TestCase` class, has this:
https://github.com/numba/numba/blob/f5af7b64b98bb745c78cf16913b16acb34017dd2/numba/tests/support.py#L42 details of the call chain follows.
    <details>

    The `pycc` module imports `numpy.distutils.misc_util` to use the external compilers bound by `numpy.distutil`. https://github.com/numba/numba/blob/f5af7b64b98bb745c78cf16913b16acb34017dd2/numba/pycc/platform.py#L6
    which is imported via:
    https://github.com/numba/numba/blob/f5af7b64b98bb745c78cf16913b16acb34017dd2/numba/pycc/cc.py#L12
    via:
    https://github.com/numba/numba/blob/f5af7b64b98bb745c78cf16913b16acb34017dd2/numba/pycc/__init__.py#L11
    which leads to any import where Numba imports something from `pycc` to trigger the use of `numpy.distutils` which then hits the problem mentioned in above.

    </details>

-----------
### Fixing it:

* Suggested short term fix is to require setuptools < 60 as suggested by the NumPy 1.23 deprecation warning.

* Suggested long term fix it to move away from `numpy.distutils`.

**Comment by esc** [comment]
#8356 has now been merged. @stuartarchibald do we leave this open as it still requires a long term fix?

**Comment by esc** [comment]
@stuartarchibald could this be a potential long-term fix: https://github.com/numba/numba/pull/8366

**Comment by stuartarchibald** [comment]
@esc #8544 should fix the issue with `numba.distutils.misc_util` as it isn't needed any more. Once implemented correctly #8476 should remove the reliance on `setuptools` at runtime.
